src/astarte_message_hub.rs

```diff
@@ -71,6 +85,182 @@ where
+            http::uri::PathAndQuery::from_static("/astarteplatform.msghub.MessageHub/Attach")
+        });
+
+        // check that the Node ID is present inside the metadata on all gRPC requests apart from the
+        // attach ones
+        if req.uri().path_and_query() == Some(&ATTACH_PATH) {
+            return self.inner.call(req).await.map_err(Into::into);
+        }
+
+        debug!("checking the NodeId in the request metadata");
+        // use tonic::Metadata to access the metadata and avoid to manually perform a base64
+        // conversion of the uuid.
+        let (mut parts, body) = req.into_parts();
+        let metadata = MetadataMap::from_headers(parts.headers);
+
+        // retrieve the node id from the metadata
+        let node_uuid = match node_id_from_bin(&metadata)? {
+            Some(node_uuid) => node_uuid,
+            None => node_id_from_ascii(&metadata)?.ok_or(InterceptorError::AbsentId)?,
+        };
+
+        trace!("NodeId: {node_uuid}");
+
+        if !self.msg_hub_nodes.read().await.contains_key(&node_uuid) {
+            debug!("NodeId not found");
+            return Err(InterceptorError::IdNotFound);
+        }
+
+        debug!("NodeId checked");
+
+        // reconstruct the http::Request
+        parts.headers = metadata.into_headers();
+        parts.extensions.insert(NodeId(node_uuid));
+        let req = http::Request::from_parts(parts, body);
+
+        self.inner.call(req).await.map_err(Into::into)
+    }
+}
+
+/// Node Id
+#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub(crate) struct NodeId(Uuid);
+
+impl Display for NodeId {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl Borrow<Uuid> for NodeId {
+    fn borrow(&self) -> &Uuid {
+        &self.0
+    }
+}
+
+impl Deref for NodeId {
+    type Target = Uuid;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+fn node_id_from_bin(metadata: &MetadataMap) -> Result<Option<Uuid>, InterceptorError> {
+    let Some(node_id) = metadata.get_bin("node-id-bin") else {
+        return Ok(None);
+    };
+
+    let uuid_bytes = node_id.to_bytes()?;
+
+    Uuid::from_slice(&uuid_bytes)
+        .map(Some)
+        .map_err(InterceptorError::Uuid)
+}
+
+fn node_id_from_ascii(metadata: &MetadataMap) -> Result<Option<Uuid>, InterceptorError> {
+    let Some(node_id) = metadata.get("node-id") else {
+        return Ok(None);
+    };
+
+    let uuid_str = node_id.to_str()?;
+
+    Uuid::from_str(uuid_str)
+        .map(Some)
+        .map_err(InterceptorError::Uuid)
+}
+
+type BoxFuture<'a, T> = Pin<Box<dyn std::future::Future<Output = T> + Send + 'a>>;
+
+impl<S> Service<hyper::Request<Body>> for NodeIdInterceptor<S>
+where
+    S: Service<hyper::Request<Body>, Response = hyper::Response<BoxBody>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: Into<InterceptorError>,
```

Comment:
I would use here the Interceptor future and Response, so we can handle the status code for the error